README.md

# Occultation Geometry Analyser
![](https://github.com/JacobParrott/OccultationProfiler/blob/master/images/Coverimage.png)
## What this project aims to achieve
In Short - It aims to create a tool to analyse occultation geometry to aid in observation planning. This tool will be built using the Spiceypy package so that the SPICE tool set( developed by NAIF ) can be employed. The first few releases of this project will aim to produce an automated Cosmographia catalogue to visualise the shape of an occultation profile ( as shown above. )

Currently this tool can analyse:
+ Solar Zenith Angle  - Describing where the sun is at the epoch of occultation ( an important aspect of atmospheric science, especially for the Martian ionosphere )
+ Distance between target and observer - So that SNR can be considered
+ Grazing Angle - To see how vertical the profile is (if the grazing angle is too large, then the top of the profile will be measuring a very different region of the planets atmosphere [could start at the poles and end on the equator.])
+ Location of the Occultations, with included map (as shown below)
![]( https://github.com/JacobParrott/Occultation-Geometry-Analyser/blob/master/images/ExampleMapMars.png )

Later on, this project will be updated to produce further details about individual occultations, such as:
+ Local times
+ Expected Rx (dB)
+ Expected Doppler


>Finally, with all the above calculated, occultations with the **perfect characteristics** can be selected via a filtering function (not implemented yet).

*This project has been written for mutual radio occultation of the Martian atmosphere. Where  radio signals are  sent from MarsExpress (MEx) to ExoMars' Trace Gas Orbiter (TGO). This can be treated as an example, where the satelites and even the planet are up to the end user's discretion. Fortunately, with the SPICE toolset, everything in the analysis can be changed with simple alterations of the starting variables.*


## Prerequisites
+ Have installed [Cosmographia 4.0](https://naif.jpl.nasa.gov/naif/cosmographia.html)
+ Some knowledge of [SPICE](https://naif.jpl.nasa.gov/naif/tutorials.html). This may take some time to understand but is very much worth it. SPICE is a remarkably powerful free-to-use tool.
+ Install the correct SPICE kernels, these are the datasets that the SPICE application reads. They include infomation on spacecraft, groundstation and planetary ephemarides, planetary shapes, onboard instrument details, leapsecond tables, reference frame conversions and *much more*. Depending on the mission the user is interested in, the download location will vary. Space agencies will manage the SPICE kernels for thier own mission. [ESA](https://www.cosmos.esa.int/web/spice) manage the kernels for [MEx](https://www.cosmos.esa.int/web/spice/spice-for-mex) and [TGO](ftp://spiftp.esac.esa.int/data/SPICE/ExoMars2016/).


### Dependancies

+ Spicypy
 + Pandas
+ Numpy
+ Matplotlib
+ Cartopy
+ OS
+ Json
+ PIL


## What is Radio Occultation?
In perticular this project will focus on mutual occultation, though the code could be easily edited to include conventional spacecraft->planatary atmosphere->earth occultations also. Mutual occultation (sometimes refered to as Cross-Link Occultatation) is a method of passing a  radiowave through an atmosphere between two spacecraft, to garner atmospheric parameters from the doppler shift exibted on the wave. This way the SNR is far better because of the proximity of the two spacecraft and due to there being no dispersive space and earth atmosphere in the radiowave's path.

## Producing a ```catalog file``` for Cosmographia
To produce and image similar to the one shown at the top of the page:
1. Achieve the prerequisites
2. Change the variables to the spacecraft, planet and date range of choice:
```
#-----------------------------------------------------<VALUES TO EDIT REGULARLY>----------------------------------------
# If you only wish to analysis mutual [cross-link] occultation between MEX and TGO, then this is the only section that
# needs to be edited
start  = '2020 JAN 1'
stop   = '2020 JAN 6'
OCCSELECTION = 7 # Which occultation do you wish to see in Cosmographia? [optional]
here = path.abspath(path.dirname(__file__))
PathtoMetaKernel1 = here + '/TGO/krns/mk/em16_plan.tm'
PathtoMetaKernel2 = here + '/MEX/krns/mk/MEX_OPS.tm'
#-----------------------------------------------------------------------------------------------------------------------


spice.furnsh(PathtoMetaKernel1)
spice.furnsh(PathtoMetaKernel2)
obs  = '-41' # NAIF code for MEX
target = '-143'# NAIF code for TGO ['EARTH'/'SUN'/ a groundstation etc]
obsfrm = 'IAU_MARS'
abcorr = 'NONE'
crdsys = 'LATITUDINAL'
coord  = 'LATITUDE'
stepsz = 300.0 # Check every 300 seconds if there is an occultation
MAXILV = 100000 #Max number of occultations that can be returned by gfoclt
bshape = 'POINT'
fshape = 'DSK/UNPRIORITIZED'
front = 'MARS'
fframe = 'IAU_MARS'
TFMT = 'YYYY-MM-DD HR:MN:SC' # Format that Cosmographia understands
```
3. This will create a Cosmographia readable .json file, now go and ``` Open Catalog... ``` and select this new Profile.json


## Feedback
Feel free file an issue. Contributors and feature requests are always welcome.


CHANGELOG.md

# Change Log
## 06/04/20
+ Residual doppler can now partially be solved, using two refractive models, one for the nuetral atmosphere and one for the ionosphere. Next to be added is a comparision between straight line rays between observer -> target and a curved refracted ray path. I am seeking a small delta in electric length between curved and straight. I want to see if I can simulate with a single straight line as this would come at a much lower computational cost. 
## 26/02/20
+ Change Log added
+ ReadMe updated to reflect new capabilities
## 25/02/20

+ Hundreds of occultations can be analysed with a single run and their geometric parameters are stored into a dataframe.
+ New geometric parameters can be calculated:
	+ Solar Zenith Angle - where the sun is at the time of occultation.
	+ Grazing Angle - angle between the Martian surface's normal and the lowest -> highest point vector.
	+ Distance - Displacement between the target and the observer.
	+ Coordinates - The latitude and longitude of the lowest tangent point (at the final epoch of occultation.)
+ Multiple locations can be plotted onto a map of mars.
+ Improvements to the Cosmographia profile former. Specifically, in 20% of cases, a maximum height would never be found and the shape of the profile would become heavily distorted. Now a max height is always found.
+ Percentage of progress through analysis is printed into the users console.
+ Code is separated into a main and setup python files to conform to best coding practice.
+ Relative file paths have been added to point to the stored kernels and pictures .

## 04/02/20
+ README file added


##  29/01/20 - Initial Commit
+ Code can produce a ```catalog file``` for Cosmographia so that the profile formed from a  single mutual radio occultation can be easily visualised.


Main.py

import numpy as np
import spiceypy as spice
import spiceypy.utils.support_types as stypes
import math
import pandas as pd
import json

#Import custom modules
import profileformer
import JSONtemplate as JS


#-----------------------------------------------------<VALUES TO EDIT REGULARLY>----------------------------------------
# If you only wish to analysis mutual [cross-link] occultation between MEX and TGO, then this is the only section that
# needs to be edited
start  = '2020 JAN 1'
stop   = '2020 JAN 6'
OCCSELECTION = 18 # Which occultation do you wish to see?
PathtoMetaKernel1 = '____'
PathtoMetaKernel2 = '_____' #[ ommit if only one mk]
#-----------------------------------------------------------------------------------------------------------------------


spice.furnsh(PathtoMetaKernel1)
spice.furnsh(PathtoMetaKernel2)
obs  = '-41' # NAIF code for MEX
target = '-143'# NAIF code for TGO ['EARTH'/'SUN'/ a groundstation etc]
obsfrm = 'IAU_MARS'
abcorr = 'NONE'
crdsys = 'LATITUDINAL'
coord  = 'LATITUDE'
stepsz = 300.0 # Check every 300 seconds if there is an occultation
MAXILV = 100000 #Max number of occultations that can be returned by gfoclt
bshape = 'POINT'
fshape = 'DSK/UNPRIORITIZED'
front = 'MARS'
fframe = 'IAU_MARS'
TFMT = 'YYYY-MM-DD HR:MN:SC' # Format that Cosmographia understands


# Setting Variables
ingresslist = np.array([1.0] ,dtype = float)
etbeg = spice.str2et( start )
etend = spice.str2et( stop  )

# Form a windows that gfoclt can populate
window = stypes.SPICEDOUBLE_CELL(2)
spice.wninsd(etbeg, etend,window)
occwindow = stypes.SPICEDOUBLE_CELL(MAXILV)

#find occultation windows between the dates listed above [ most comp cost in this function]
spice.gfoclt('ANY',front,fshape,fframe, target, bshape, 'J2000' , abcorr, obs, stepsz, window, occwindow)

winsiz = spice.wncard( occwindow )# Find cardinality (number of windows)

# Inter the ingress epochs into a dataframe
occlist = np.ones((winsiz,3))
for i in range(winsiz):
    [ingress, eegress] = spice.wnfetd(occwindow, i) # extract the begining and ends of the windows
    if i == 1 :
        ingresslist = ingress
    else:
        ingresslist = np.append(ingresslist, [ingress])
occ=np.transpose(range(winsiz-1))
df = pd.DataFrame(ingresslist, columns=['Time'])

# Select and occultation of interest, calculate the shape of it's profile, then add this to a dateframe
et = df.Time[OCCSELECTION]
result = profileformer.occgradient(front,et , fframe,obs,target )
profile = result[0]
lowesttime = result[1]
highesttime = result[2]
lowesttimeFMT = spice.timout((et-result[1]),TFMT)
highesttimeFMT = spice.timout((et-result[2]),TFMT)
endtimeFMT = spice.timout(et,TFMT)
profile = np.transpose(profile)
profiledataframe = pd.DataFrame(profile[:][0:highesttime], columns = ['X', 'Y', 'Z'])

# Example polygon velocities [this effects the dynamic texture of the profile, purely aesthetic]. These values have
# been chosen to be slow and variable
vs = np.array([[1.4236, -2.4657, 6.3948],
            [1.6404, -2.2997, 6.4047],
            [1.8386, -2.1150, 6.4145],
            [2.0166, -1.9136, 6.4243],
            [2.1730, -1.6977, 6.4339],
            [2.3068, -1.4696, 6.4435],
            [2.4170, -1.2315, 6.4530],
            [2.5029, -0.9859, 6.4625],
            [2.5029, -0.9859, 6.4625],
            [2.5029, -0.9859, 6.4625]]) 


# Iterate the polygon velocity states to be the length of profile and then combine with the positions
vs = vs.repeat(200,axis=0)
vt = vs[:][0:highesttime]
velocitiesdataframe = pd.DataFrame(vt, columns = ['dX', 'dY', 'dZ'])
finalprofile = pd.concat([profiledataframe,velocitiesdataframe],axis=1)


# Construct a JSON template depending on the size of the profile, split into sections of 10 points [smaller sections,
# more smoothly the profile forms over time]
blockcount = math.floor(highesttime/10)*10
i=0
while i < (blockcount/10):
  JS.JSONiterated = JS.JSONiterated +JS.JSONiterable
  i=i+1
JSONtemplate = JS.JSONstart + JS.JSONiterated + JS.JSONend


template = json.loads(JSONtemplate) #load the template created above  so it can be editted as JSON

itemcounter = 0

#convert states into single string for cosmographia to understand [cosmogrpahia computes as
# 3 postions and 3 velocities, no need for '\n']
for i in range(0,blockcount,10):
  block = (finalprofile[i:i+10].to_numpy())*(-1) #inverse due to reference frame inversion
  ProfileBlock = block.tolist() # convert to list for the JSON format

  states=[]
  for p in range(10):
    states.extend(ProfileBlock[p]) #turn states into one long line to remove the '\n', which confuses cosmographia

    # vary the spread and intensity of the profile as it increases with alitude and the atmosphere becomes less dense
  profilespread = math.floor(itemcounter * (5/(blockcount/10)))
  profileintensity = math.floor(itemcounter * (30 / (blockcount / 10)))

  # edit the large JSON file iterativly, with 'itemcounter'moving down the JSON items (blocks of 10 profile points).
  # adding the states(position and velocity of the polygons), name of item (so comsmographia doesnt overwrite them),
  # the start time, end time(epoch of occultation), the spread and finally the intensity of colour
  template['items'][itemcounter]['geometry']['emitters'][0]['generator']['states'] = states
  template['items'][itemcounter]['name'] = 'Profile ' + str(itemcounter)
  template['items'][itemcounter]['geometry']['emitters'][0]['startTime']  = spice.timout((et-i),TFMT)
  template['items'][itemcounter]['geometry']['emitters'][0]['endTime'] = endtimeFMT
  template['items'][itemcounter]['geometry']['emitters'][0]['velocityVariation'] = (profilespread +1)
  template['items'][itemcounter]['geometry']['emitters'][0]['endSize'] = profileintensity
  itemcounter=itemcounter+1

#serialise the formed profile into a json that cosmographia can read
with open( ' Profile.json'  , 'w') as file:
    json.dump(template,file, indent = 3)










